coreax/solvers/recombination.py

```diff
@@ -0,0 +1,570 @@
+        safe_push_forward_nodes, safe_weights, indices = _co_linearize(
+            push_forward_nodes, weights
+        )
+        largest_null_space_basis, null_space_rank = _resolve_null_basis(
+            safe_push_forward_nodes, self.rcond
+        )
+
+        def _eliminate_cond(state: _EliminationState) -> bool:
+            """
+            If to continue the iterative Gaussian-Elimination procedure.
+
+            On each iteration, we eliminate a basis vector from the left null space. We
+            repeat until all basis vectors have been eliminated (the dimension of the
+            null space is zero); once the number of iterations is the same as the rank
+            of the original null space.
+
+            .. note::
+                The reason for using a while loop, rather than scanning over the basis
+                vectors, is due to the dimension of the null space being unknown at JIT
+                compile time, preventing us from slicing the left singular vectors down
+                to only those which form a basis for the left null space.
+            """
+            *_, basis_index = state
+            return basis_index < null_space_rank
+
+        def _eliminate(state: _EliminationState) -> _EliminationState:
+            """
+            Eliminate a basis from the left null space.
+
+            At least one weight is zeroed (implicitly removed from the dataset), and one
+            left null space basis vector eliminated on each iteration. The mass that is
+            "lost" in weight zeroing/elimination is redistributed among the remaining
+            non-zero weights to preserve the total mass/weight sum.
+
+            If the procedure is repeated until all the left null space basis vectors
+            are eliminated, the resulting weights (when combined with the original
+            nodes) are a BFS to the recombination problem/linear-program.
+            """
+            _weights, null_space_basis, basis_index = state
+            basis_vector = null_space_basis[basis_index]
+            _elimination_condition = _weights / basis_vector
+            elimination_condition = jnp.where(
+                basis_vector > 0, _elimination_condition, jnp.inf
+            )
+            elimination_index = jnp.argmin(elimination_condition)
+            elimination_rescaling_factor = elimination_condition[elimination_index]
+            updated_weights = _weights - elimination_rescaling_factor * basis_vector
+            updated_weights = updated_weights.at[elimination_index].set(0)
+            rescaled_basis_vector = basis_vector / basis_vector[elimination_index]
+            null_space_basis_update = jnp.tensordot(
+                null_space_basis[:, elimination_index], rescaled_basis_vector, axes=0
+            )
+            updated_null_space_basis = null_space_basis - null_space_basis_update
+            updated_null_space_basis = updated_null_space_basis.at[basis_index].set(0)
```

Comment:
Some more comments here detailing what each step is doing mathematically would be really helpful

Reply:
Appologies for saying this again, but please you could be more specific here on what specifically you find to be unclear.

Reply:
I think this is just quite difficult to read in its current form, separating it into blocks with some minimal comment explaining the purpose of each block would make reviewing (and comparing to the paper) easier. Any equations/lines in algorithms you can refer to would make this easier too